aptos-indexer-processors-sdk/sdk/src/common_steps/insertion_limit_step.rs

```diff
@@ -0,0 +1,298 @@
+use crate::{
+    traits::{AsyncRunType, AsyncStep, NamedStep, Processable},
+    types::transaction_context::TransactionContext,
+    utils::{
+        errors::ProcessorError,
+        step_metrics::{StepMetricLabels, INSERTION_LIMIT_STEP_REMAINING_BYTES},
+    },
+};
+use serde::{Deserialize, Serialize};
+use std::{
+    marker::PhantomData,
+    time::{Duration, Instant},
+};
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct InsertionLimitConfig {
+    /// This is the maximum number of bytes that can be inserted in the given number of seconds.
+    pub num_bytes: u64,
+    /// This is the number of seconds over which the `num_bytes` limit is applied.
+    pub num_seconds: u64,
+}
+
+/// This step limits the number of bytes that can be inserted per second, based on a
+/// config specifying the number of bytes that can be inserted in a given number of
+/// seconds. If we hit the limit, we sleep until we know there will be enough tokens to
+/// insert the items, at which point we pass the item on to the next step.
+///
+/// We mandate that the Input implement Sizeable so we can calculate its size. The dev
+/// can implement this whichever way works best for them, e.g. making their struct
+/// derive allocative::Allocative or get_size::GetSize and then using the result from
+/// the related functions, or just implementing it by hand.
+///
+/// This should go before the DB writing step.
+pub struct InsertionLimitStep<Input>
+where
+    Self: Sized + Send + 'static,
+    Input: Send + Sizeable + 'static,
+{
+    /// This is the maximum number of bytes that can be inserted in the given number of seconds.
+    capacity: f64,
+    /// This is the current number of tokens (the number of bytes that can be inserted this instant).
+    tokens: f64,
```

Comment:
maybe rename capacity to like `max_insertions_per_bucket` and tokens to `bytes_available_in_current_bucket` or something idk